File: Embarcadero/Seattle/Object Pascal/Bluetooth/Beacons/Beacon Device/UBeaconDevDemo.pas
unit UBeaconDevDemo;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs, System.Beacon, System.Beacon.Components, FMX.StdCtrls,
  FMX.Controls.Presentation, FMX.Ani, FMX.Objects, FMX.Edit;

type
  TForm37 = class(TForm)
    EdtBeaconUUID: TEdit;
    EdtBeaconMajor: TEdit;
    EdtBeaconMinor: TEdit;
    Label3: TLabel;
    Label4: TLabel;
    Label5: TLabel;
    ImageControl1: TImageControl;
    Animation: TBitmapListAnimation;
    StyleBook1: TStyleBook;
    PnlBeaconInfo: TPanel;
    Label1: TLabel;
    Panel2: TPanel;
    BtnRandom: TButton;
    Label6: TLabel;
    EdTxPower: TEdit;
    BtnEnableBeacon: TButton;
    BeaconDevice1: TBeaconDevice;
    procedure BtnRandomClick(Sender: TObject);
    procedure BtnEnableBeaconClick(Sender: TObject);
  private
    { Private declarations }
    FGuid: TGUID;
    FMajor: Integer;
    FMinor: Integer;
    FTxPower: Integer;
    function CheckValues: Boolean;
  public
    { Public declarations }
  end;

var
  Form37: TForm37;

implementation

{$R *.fmx}

procedure TForm37.BtnRandomClick(Sender: TObject);
var
  LGuid: TGUID;
begin
  CreateGUID(LGuid);
  EdtBeaconUUID.Text := Copy(LGuid.ToString, 2, Length(LGuid.ToString) - 2);
  EdtBeaconMajor.Text := IntToStr(Random(High(Word)));
  EdtBeaconMinor.Text := IntToStr(Random(High(Word)));
end;

procedure TForm37.BtnEnableBeaconClick(Sender: TObject);
begin
  if BtnEnableBeacon.IsPressed then
    if CheckValues then
    begin
      BeaconDevice1.GUID := FGuid;
      BeaconDevice1.Major := FMajor;
      BeaconDevice1.Minor := FMinor;
      BeaconDevice1.TxPower := FTxPower;
    end
    else
      BtnEnableBeacon.IsPressed := False;

  if BtnEnableBeacon.IsPressed then
    BtnEnableBeacon.Text := 'Disable Beacon'
  else
    BtnEnableBeacon.Text := 'Enable Beacon';

  BeaconDevice1.Enabled := BtnEnableBeacon.IsPressed;
  ImageControl1.Visible := BeaconDevice1.Enabled;
  Animation.Enabled := BeaconDevice1.Enabled;
  PnlBeaconInfo.Enabled := not BeaconDevice1.Enabled;
end;

function TForm37.CheckValues: Boolean;
begin
  try
    FGuid := StringToGUID('{' + EdtBeaconUUID.Text + '}');
  except
    on EConvertError do
    begin
      ShowMessage(EdtBeaconUUID.Text + ' is not a valid UUID value');
      Exit(False);
    end;
  end;
  FMajor := StrToIntDef(EdtBeaconMajor.Text, 0);
  EdtBeaconMajor.Text := IntToStr(FMajor);
  FMinor := StrToIntDef(EdtBeaconMinor.Text, 0);
  EdtBeaconMinor.Text := IntToStr(FMinor);
  FTxPower := StrToIntDef(EdTxPower.Text, -56);
  EdTxPower.Text := IntToStr(FTxPower);
  Result := True;
end;

end.

File: Embarcadero/Seattle/CPP/Beacons/Beacon Device/UBeaconDevDemo.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------
//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "UBeaconDevDemo.h"
#include <limits.h>
#include <string.h>

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TForm8 *Form8;
//---------------------------------------------------------------------------
__fastcall TForm8::TForm8(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
bool __fastcall TForm8::CheckValues()
{
	try
	{
		FGuid = StringToGUID("{" + EdtBeaconUUID->Text + "}");
	}
	catch (EConvertError &e)
	{
	  ShowMessage(EdtBeaconUUID->Text + " is not a valid UUID value");
	  return False;
	}
	FMajor = StrToIntDef(EdtBeaconMajor->Text, 0);
	EdtBeaconMajor->Text = IntToStr(FMajor);
	FMinor = StrToIntDef(EdtBeaconMinor->Text, 0);
	EdtBeaconMinor->Text = IntToStr(FMinor);
	FTxPower = StrToIntDef(EdTxPower->Text, -56);
	EdTxPower->Text = IntToStr(FTxPower);
	return True;
}
//---------------------------------------------------------------------------
void __fastcall TForm8::BtnRandomClick(TObject *Sender)
{
	TGUID LGuid;
	CreateGUID(LGuid);
	UnicodeString LStrGuid = GUIDToString(LGuid);
	EdtBeaconUUID->Text = LStrGuid.SubString0(1, LStrGuid.Length() - 2);
	EdtBeaconMajor->Text = IntToStr(Random(USHRT_MAX));
	EdtBeaconMinor->Text = IntToStr(Random(USHRT_MAX));
}

//---------------------------------------------------------------------------
void __fastcall TForm8::BtnEnableBeaconClick(TObject *Sender) {
	if (BtnEnableBeacon->IsPressed)
	{
		if (CheckValues()) {
			BeaconDevice1->GUID = FGuid;
			BeaconDevice1->Major = FMajor;
			BeaconDevice1->Minor = FMinor;
			BeaconDevice1->TxPower = FTxPower;
		}
		else
		{
			BtnEnableBeacon->IsPressed = False;
		}
	}

	if (BtnEnableBeacon->IsPressed)
		BtnEnableBeacon->Text = "Disable Beacon";
	else
		BtnEnableBeacon->Text = "Enable Beacon";

	BeaconDevice1->Enabled = BtnEnableBeacon->IsPressed;
	ImageControl1->Visible = BeaconDevice1->Enabled;
	Animation->Enabled = BeaconDevice1->Enabled;
	PnlBeaconInfo->Enabled = !BeaconDevice1->Enabled;
}
//---------------------------------------------------------------------------
